include/eve/algo/container/soa_vector2.hpp

```diff
@@ -0,0 +1,372 @@
+      {
+        eve::write(*ptr++, begin()+i);
+      }
+    }
+
+    //==============================================================================================
+    //! @}
+    //==============================================================================================
+
+    //==============================================================================================
+    //! @name Capacity
+    //! @{
+    //==============================================================================================
+    //! Returns the number of elements in the container
+    EVE_FORCEINLINE std::size_t size()  const noexcept { return storage.size_; }
+
+    //! Returns the number of elements that the container has currently allocated space for
+    EVE_FORCEINLINE std::size_t capacity()  const noexcept { return storage.capacity_; }
+
+    //! Requests the removal of unused capacity.
+    // EVE_FORCEINLINE void shrink_to_fit() noexcept
+    // {
+    //   kumi::for_each([](auto& m) { m.shrink_to_fit(); }, storage);
+    // }
+
+    //! Checks if the container has no elements
+    EVE_FORCEINLINE bool empty() const noexcept { return size() == 0ULL; }
+
+    //! @brief Increase the capacity of the vector to a value that's greater or equal to `n`.
+    //! If `n` is greater than the current capacity(), new storage is allocated and all iterators,
+    //! including the past-the-end iterator, and all references to the elements are invalidated.
+    //! Otherwise, the method does nothing.
+    //! @warning reserve() does not change the size of the vector.
+    //! @param n New capacity of the `eve::soa_vector`
+    void reserve(std::size_t n)
+    {
+      if(n > capacity())
+      {
+        auto const sz = size();
+        soa_vector that(n);
+        eve::algo::copy ( *this
+                        , algo::as_range(that.begin(), that.begin() + sz)
+                        );
+        this->operator=(std::move(that));
+        storage.size_ = sz;
+     }
+    }
+
+    //==============================================================================================
+    //! @}
+    //==============================================================================================
+
+    //==============================================================================================
+    //! @name Modifiers
+    //! @{
+    //==============================================================================================
+    //! @brief Clear the contents of the container.
+    void clear() { storage.size_ = 0; }
+
+    //! @brief Removes an element from the container.
+    //! Has the same invalidation semantics as std::vector.
+    //! end() iterator is not a valid pos.
+    iterator erase(const_iterator pos)
+    {
+      std::ptrdiff_t distance = pos - cbegin();
+      eve::algo::copy_backward( eve::algo::as_range(pos+1, cend())
+                              , eve::algo::as_range(begin()+distance, begin() + size() - 1)
```

Comment:
you can also

```
eve::algo::copy_backward(eve::algo::zip(
      eve::algo::as_range(pos+1, cend()), 
      begin()+distance));
```

We just require people to zip explicitly to avoid difference in semantics with the standard.

You can also have a version with `const_iterator_aligned` and then figure out how to use next aligned after cend. Let's not now and do in a separate patch though.